lib/events.js

```diff
@@ -647,3 +652,84 @@ function once(emitter, name) {
     emitter.once(name, eventListener);
   });
 }
+
+const AsyncIteratorPrototype = ObjectGetPrototypeOf(
+  ObjectGetPrototypeOf(async function* () {}).prototype);
+
+function createIterResult(value, done) {
+  return { value, done };
+}
+
+function on(emitter, event) {
+  const unconsumedEvents = [];
+  const unconsumedPromises = [];
+  let error = null;
+
+  const iterator = ObjectSetPrototypeOf({
+    next() {
+      if (error) {
```

Comment:
This leaks `unconsumedEvents` ;

If a emitter emits three data events & one error synchronously and you have only called next once then theres two pending data & 1 pending error.

The following `next()` call will yield the error and the two pending data events are just sitting in `unconsumedEvents`

Reply:
That is by design. The iterator will go out of scope and everything will be cleaned up.

Reply:
I did not mean it leaking memory. I meant it swallowing data.

It’s zalgo behavior for it to return the error and events out of order. The error should really be pushed onto the unconsumedEvents array so it maintains the order in which values were emitted on the event emitters

Reply:
Fixed.